exceptions refactor part 2

saneshark opened the pull request.

saneshark commented:
@aub this is still just Phase 2 of the refactor. I wanted to refactor the common exceptions to use yaml locales with interpolation as this will be a necessary step for the BackendServerError.

I have deliberately tried to keep the DSL for the exception classes consistent with how they've been. Because I'm not 100% certain yet that there is adequate end-to-end test coverage on them. My changes to the specs are minimal.

The next phase, I plan to move the application_controller rescue_from logic into a Common::ExceptionHandler class. 

Make ClientError accept an exception instead of attributes (exception must be of type BackendServerError)

Make BackendServerError accept arguments, reimplement the raise_error middleware to properly build the arguments and raise the exception for MHV, or alternately make it an adapter and have it accept arguments for its own middleware / parsing mechanism that is service specific.

aub commented on app/controllers/application_controller.rb:
```diff
@@ -37,8 +37,12 @@ def routing_error
+          source: exception.developer_message,
+          meta: exception.as_json
+        }
+        Common::Exceptions::ClientError.new(nil, additional_attributes)
```
I guess I'm still a little bit confused here, but:

* Shouldn't this error also be renamed to indicate a problem with the backend service?
* Why do we have to make this translation here? Couldn't we just raise this exception type in the first place?
* I'm sure there's a semantic difference between the classes under `Common::Client::Errors` and `Common::Exceptions`, but I'm not sure the naming makes it clear what that difference is.

saneshark replied:
Yes, I can rename it. Would Common::Exceptions::BackendServiceError be more appropriate? 

I think I'd prefer to keep client exceptions distinct from common exceptions for now. The client has its own distinct responsibilities, and I don't think it should be concerned with having the capability of rendering according to JSONAPI spec. I think that is a distinctly separate responsibility that Common::Exceptions should handle.

I think I agree with you though. To me the responsibilities for everything under Client are distinct from Common::Exceptions. They don't need to be, but I think as it stands now, were these to be packaged as a gem, I would consider them as 2 separate gems.

Hope that helps to clarify. Let me know if it doesn't.

aub replied:
Sure, BackendServiceError makes sense.

I think the problem I'm having with `Common::Client::Errors` and `Common::Exceptions` is that I still don't quite understand what they're intended to do. I think I'm starting to see that `Common::Client::Errors` is meant to handle errors arising from communication with the backend APIs while `Common::Exceptions` manage issues coming from the Rails app itself. If that's the case, then maybe the problem is solved by replacing the confusing word `Client` with something more specific to the problem these classes are addressing?

saneshark commented:
@aub I think maybe it would be wise for me to build a linter of sorts for the locales file. To ensure codes are unique, that the required fields are provided. I'm not sure if this is best to do as rspec or as part of an initializer. If this were a gem, I would think it would be in an initializer.

aub commented:
Could you write that as a rake task and have that task be amongst the things that get run during CI?

saneshark commented:
I will add it to my list of TODO items. I don't think its necessary for this PR, and would like to maybe make the other changes I described above before working on the linter, but I like the idea of having a rake task that can be triggered by CI.

aub commented:
Yep, agreed that it can be a separate PR.

saneshark commented:
@aub I changed the name of the common exceptions class. I also added some helpful README files, perhaps having a base README that links to the individual common components. I'd like to also write one up for the models folder. I think I will organize the README files later, but I thought these would be helpful for now based on conversation you and I have had in this thread.

saneshark commented:
@patrickvinograd could you take a peak at this line in multi_client? I know you have a PR open to make changes there, but I wanted to make sure this is going to work for you in the interim.

https://github.com/department-of-veterans-affairs/vets-api/pull/568/files#diff-cfe4dff4bda116c6c4d62961ee4d6831L19

carlmjohnson commented on app/controllers/application_controller.rb:
```diff
@@ -44,8 +44,12 @@ def routing_error
+          source: exception.developer_message,
+          meta: exception.as_json
+        }
+        Common::Exceptions::BackendServiceError.new(nil, additional_attributes)
```
I find it a bit confusing that we're converting a `BackendServiceError` to a `BackendServiceError`. Different namespaces, of course, but I don't think that will necessarily be obvious to future devs.

saneshark replied:
Haha, sounds like someone is going to be confused one way or the other on this one. @carlmjohnson this change was mostly meant to address concerns @aub brought up on the client_error not being clear enough -- I'm hoping that eventually one or the other of these will go away entirely, but I can't think of a good way to facilitate this yet. Maybe once I've had to revisit the one on client side in Phase 3 of this refactor.

carlmjohnson commented on lib/common/exceptions/external/backend_service_error.rb:
```diff
@@ -0,0 +1,24 @@
+module Common
+  module Exceptions
+    # ClientError - Generic Backend errors returned from api calls
+    class BackendServiceError < BaseError
```
What about calling this one a `BackendServiceException`?

patrickvinograd commented:
LGTM.
